File: mobile/lib/core/storage/intent_queue.dart
/// Intent Queue — SQLite-backed pending-operation recovery for Flutter
///
/// Mirrors the web IndexedDB intent queue. On app startup, replays any
/// PENDING intents using the same UUID as the X-Idempotency-Key header,
/// which triggers the backend's idempotency cache and returns the cached
/// result rather than executing the handler a second time.
///
/// Dependencies:
///   sqflite: ^2.3.2
///   path: ^1.9.0
///
/// Add to pubspec.yaml:
///   sqflite: ^2.3.2
///   path: ^1.9.0

import 'dart:convert';
import 'package:sqflite/sqflite.dart';
import 'package:path/path.dart' as p;
import 'package:uuid/uuid.dart';

// ─── Types ────────────────────────────────────────────────────────────────────

enum IntentType {
  issueCredential,
  generateProof,
  submitPresentation,
  registerUser,
  recoverAccount,
}

enum IntentStatus { pending, completed, failed }

class Intent {
  final String       id;             // UUID — also the idempotency key
  final IntentType   type;
  final Map<String, dynamic> payload;
  final IntentStatus status;
  final int          createdAt;      // epoch ms
  final int          attempts;
  final dynamic      result;
  final String?      error;

  const Intent({
    required this.id,
    required this.type,
    required this.payload,
    required this.status,
    required this.createdAt,
    required this.attempts,
    this.result,
    this.error,
  });

  Map<String, dynamic> toRow() => {
    'id':         id,
    'type':       type.name,
    'payload':    jsonEncode(payload),
    'status':     status.name,
    'created_at': createdAt,
    'attempts':   attempts,
    'result':     result != null ? jsonEncode(result) : null,
    'error':      error,
  };

  factory Intent.fromRow(Map<String, dynamic> row) => Intent(
    id:         row['id'] as String,
    type:       IntentType.values.firstWhere((t) => t.name == row['type']),
    payload:    jsonDecode(row['payload'] as String) as Map<String, dynamic>,
    status:     IntentStatus.values.firstWhere((s) => s.name == row['status']),
    createdAt:  row['created_at'] as int,
    attempts:   row['attempts'] as int,
    result:     row['result'] != null ? jsonDecode(row['result'] as String) : null,
    error:      row['error'] as String?,
  );
}

// ─── Queue ────────────────────────────────────────────────────────────────────

class IntentQueueService {
  static const int _maxAttempts = 5;
  static const String _dbName   = 'zk_auth_intents.db';
  static const int    _version  = 1;

  Database? _db;
  final _uuid = const Uuid();

  // ─── Lifecycle ─────────────────────────────────────────────────────────────

  Future<void> init() async {
    final dbPath = p.join(await getDatabasesPath(), _dbName);
    _db = await openDatabase(
      dbPath,
      version: _version,
      onCreate: (db, version) async {
        await db.execute('''
          CREATE TABLE intents (
            id         TEXT PRIMARY KEY,
            type       TEXT NOT NULL,
            payload    TEXT NOT NULL,
            status     TEXT NOT NULL DEFAULT 'pending',
            created_at INTEGER NOT NULL,
            attempts   INTEGER NOT NULL DEFAULT 0,
            result     TEXT,
            error      TEXT
          )
        ''');
        await db.execute(
          'CREATE INDEX idx_intents_status ON intents(status)',
        );
      },
    );
  }

  // ─── Public API ────────────────────────────────────────────────────────────

  /// Persist a new PENDING intent. Returns the UUID (idempotency key).
  Future<String> enqueue(
    IntentType               type,
    Map<String, dynamic>     payload,
  ) async {
    _assertOpen();
    final intent = Intent(
      id:        _uuid.v4(),
      type:      type,
      payload:   payload,
      status:    IntentStatus.pending,
      createdAt: DateTime.now().millisecondsSinceEpoch,
      attempts:  0,
    );
    await _db!.insert('intents', intent.toRow());
    return intent.id;
  }

  /// Mark an intent as COMPLETED with the API result.
  Future<void> complete(String id, dynamic result) async {
    _assertOpen();
    await _db!.update(
      'intents',
      {
        'status': 'completed',
        'result': result != null ? jsonEncode(result) : null,
      },
      where: 'id = ?',
      whereArgs: [id],
    );
  }

  /// Increment attempt count; mark as FAILED after maxAttempts.
  Future<void> fail(String id, String error) async {
    _assertOpen();
    final rows = await _db!.query(
      'intents', columns: ['attempts'], where: 'id = ?', whereArgs: [id],
    );
    if (rows.isEmpty) return;

    final attempts = (rows.first['attempts'] as int) + 1;
    await _db!.update(
      'intents',
      {
        'attempts': attempts,
        'error':    error,
        'status':   attempts >= _maxAttempts ? 'failed' : 'pending',
      },
      where: 'id = ?',
      whereArgs: [id],
    );
  }

  /// Return all PENDING intents for replay on startup.
  Future<List<Intent>> getPending() async {
    _assertOpen();
    final rows = await _db!.query(
      'intents', where: 'status = ?', whereArgs: ['pending'],
      orderBy: 'created_at ASC',
    );
    return rows.map(Intent.fromRow).toList();
  }

  /// Replay all PENDING intents.
  ///
  /// [handler] receives the intent and should call the API with
  /// `intent.id` as the `X-Idempotency-Key` header.
  Future<({int replayed, int failed})> replayPending(
    Future<dynamic> Function(Intent) handler,
  ) async {
    final pending = await getPending();
    int replayed  = 0;
    int failedCount = 0;

    for (final intent in pending) {
      try {
        final result = await handler(intent);
        await complete(intent.id, result);
        replayed++;
      } catch (e) {
        await fail(intent.id, e.toString());
        failedCount++;
      }
    }

    return (replayed: replayed, failed: failedCount);
  }

  /// Delete COMPLETED intents older than [maxAge] (default 7 days).
  Future<int> prune([Duration maxAge = const Duration(days: 7)]) async {
    _assertOpen();
    final cutoff = DateTime.now()
        .subtract(maxAge)
        .millisecondsSinceEpoch;

    return _db!.delete(
      'intents',
      where: 'status = ? AND created_at < ?',
      whereArgs: ['completed', cutoff],
    );
  }

  Future<void> dispose() async => _db?.close();

  void _assertOpen() {
    if (_db == null) throw StateError('IntentQueueService not initialised — call init() first');
  }
}

/// Global singleton
final intentQueue = IntentQueueService();

File: web/src/lib/intentQueue.ts
/**
 * Intent Queue — IndexedDB-backed pending-operation recovery
 *
 * Solves the force-quit problem:
 *   1. Before calling any mutating API, persist an Intent record locally.
 *   2. On every app startup, the queue worker scans for PENDING intents.
 *   3. Pending intents are replayed with their original idempotency key.
 *   4. The backend's idempotency middleware returns the cached result if
 *      the operation already completed — no duplicate side-effects.
 *
 * ─── IndexedDB schema ─────────────────────────────────────────────────────────
 *   DB:    zk_auth_intents   version: 1
 *   Store: intents           keyPath: id
 *     id:         UUID (also used as X-Idempotency-Key)
 *     type:       IntentType
 *     payload:    Record<string, unknown>
 *     status:     'PENDING' | 'COMPLETED' | 'FAILED'
 *     createdAt:  number (epoch ms)
 *     attempts:   number
 *     result?:    unknown
 *     error?:     string
 *
 * ─── Usage ────────────────────────────────────────────────────────────────────
 *
 *   // Before the API call:
 *   const intentId = await intentQueue.enqueue('ISSUE_CREDENTIAL', { userId, ... });
 *
 *   // Make the API call:
 *   const result = await api.post('/api/issuer/issue-id', body, {
 *     headers: { 'X-Idempotency-Key': intentId }
 *   });
 *
 *   // On success:
 *   await intentQueue.complete(intentId, result);
 *
 *   // On failure (or app crash — the intent stays PENDING):
 *   await intentQueue.fail(intentId, error.message);
 *
 *   // On app start — replay:
 *   await intentQueue.replayPending(handler);
 */

export type IntentType =
  | 'ISSUE_CREDENTIAL'
  | 'GENERATE_PROOF'
  | 'SUBMIT_PRESENTATION'
  | 'REGISTER_USER'
  | 'RECOVER_ACCOUNT';

export type IntentStatus = 'PENDING' | 'COMPLETED' | 'FAILED';

export interface Intent<T = Record<string, unknown>> {
  id:        string;         // UUID v4 — also the idempotency key
  type:      IntentType;
  payload:   T;
  status:    IntentStatus;
  createdAt: number;
  attempts:  number;
  result?:   unknown;
  error?:    string;
}

const DB_NAME    = 'zk_auth_intents';
const DB_VERSION = 1;
const STORE_NAME = 'intents';
const MAX_ATTEMPTS = 5;

// ─── DB open helper ───────────────────────────────────────────────────────────

function openDB(): Promise<IDBDatabase> {
  return new Promise((resolve, reject) => {
    const req = indexedDB.open(DB_NAME, DB_VERSION);

    req.onupgradeneeded = (e) => {
      const db = (e.target as IDBOpenDBRequest).result;
      if (!db.objectStoreNames.contains(STORE_NAME)) {
        const store = db.createObjectStore(STORE_NAME, { keyPath: 'id' });
        store.createIndex('status',    'status',    { unique: false });
        store.createIndex('createdAt', 'createdAt', { unique: false });
      }
    };

    req.onsuccess = () => resolve(req.result);
    req.onerror   = () => reject(req.error);
  });
}

function txPromise<T>(
  db:   IDBDatabase,
  mode: IDBTransactionMode,
  fn:   (store: IDBObjectStore) => IDBRequest<T>,
): Promise<T> {
  return new Promise((resolve, reject) => {
    const tx    = db.transaction(STORE_NAME, mode);
    const store = tx.objectStore(STORE_NAME);
    const req   = fn(store);

    req.onsuccess = () => resolve(req.result);
    req.onerror   = () => reject(req.error);
    tx.onerror    = () => reject(tx.error);
  });
}

// ─── IntentQueue ──────────────────────────────────────────────────────────────

export class IntentQueue {
  private _db: IDBDatabase | null = null;

  async init(): Promise<void> {
    if (typeof window === 'undefined' || !window.indexedDB) return;
    this._db = await openDB();
  }

  /** Persist a new intent and return its UUID (used as idempotency key). */
  async enqueue<T extends Record<string, unknown>>(
    type:    IntentType,
    payload: T,
  ): Promise<string> {
    if (!this._db) return crypto.randomUUID();

    const intent: Intent<T> = {
      id:        crypto.randomUUID(),
      type,
      payload,
      status:    'PENDING',
      createdAt: Date.now(),
      attempts:  0,
    };

    await txPromise(this._db, 'readwrite', (store) => store.add(intent));
    return intent.id;
  }

  /** Mark an intent as successfully completed, storing the result. */
  async complete(id: string, result: unknown): Promise<void> {
    if (!this._db) return;
    await this._patch(id, { status: 'COMPLETED', result });
  }

  /** Increment attempt count and optionally mark as FAILED. */
  async fail(id: string, error: string): Promise<void> {
    if (!this._db) return;
    const existing = await txPromise<Intent>(
      this._db, 'readonly', (store) => store.get(id),
    );
    if (!existing) return;

    const newAttempts = existing.attempts + 1;
    await this._patch(id, {
      attempts: newAttempts,
      error,
      status: newAttempts >= MAX_ATTEMPTS ? 'FAILED' : 'PENDING',
    });
  }

  /** Retrieve all PENDING intents for replay on app startup. */
  async getPending(): Promise<Intent[]> {
    if (!this._db) return [];
    return new Promise((resolve, reject) => {
      const tx    = this._db!.transaction(STORE_NAME, 'readonly');
      const store = tx.objectStore(STORE_NAME);
      const index = store.index('status');
      const req   = index.getAll('PENDING');
      req.onsuccess = () => resolve(req.result as Intent[]);
      req.onerror   = () => reject(req.error);
    });
  }

  /**
   * Replay all PENDING intents.
   *
   * @param handler — async function that receives an intent and should
   *                  call the API with intent.id as the idempotency key.
   *                  Returns the API result on success or throws on error.
   */
  async replayPending(
    handler: (intent: Intent) => Promise<unknown>,
  ): Promise<{ replayed: number; failed: number }> {
    const pending = await this.getPending();
    let replayed  = 0;
    let failed    = 0;

    for (const intent of pending) {
      try {
        const result = await handler(intent);
        await this.complete(intent.id, result);
        replayed++;
      } catch (err) {
        const msg = err instanceof Error ? err.message : String(err);
        await this.fail(intent.id, msg);
        failed++;
      }
    }

    return { replayed, failed };
  }

  /** Delete COMPLETED intents older than maxAgeMs (default 7 days). */
  async prune(maxAgeMs = 7 * 24 * 60 * 60 * 1_000): Promise<number> {
    if (!this._db) return 0;
    const cutoff = Date.now() - maxAgeMs;

    return new Promise((resolve, reject) => {
      const tx    = this._db!.transaction(STORE_NAME, 'readwrite');
      const store = tx.objectStore(STORE_NAME);
      const index = store.index('createdAt');
      const range = IDBKeyRange.upperBound(cutoff);
      const req   = index.openCursor(range);
      let count   = 0;

      req.onsuccess = (e) => {
        const cursor = (e.target as IDBRequest<IDBCursorWithValue>).result;
        if (cursor) {
          const intent = cursor.value as Intent;
          if (intent.status === 'COMPLETED') {
            cursor.delete();
            count++;
          }
          cursor.continue();
        } else {
          resolve(count);
        }
      };
      req.onerror = () => reject(req.error);
    });
  }

  // ─── Private ──────────────────────────────────────────────────────────────

  private async _patch(id: string, update: Partial<Intent>): Promise<void> {
    if (!this._db) return;
    const existing = await txPromise<Intent>(
      this._db, 'readonly', (store) => store.get(id),
    );
    if (!existing) return;

    await txPromise(
      this._db,
      'readwrite',
      (store) => store.put({ ...existing, ...update }),
    );
  }
}

export const intentQueue = new IntentQueue();
